src/impression.js

```diff
@@ -53,74 +53,119 @@ export function resetTrackImpressionPromiseForTesting() {
 export function maybeTrackImpression(win) {
   let resolveImpression;
 
-  trackImpressionPromise = new Promise(resolve => {
+  const promise = new Promise(resolve => {
     resolveImpression = resolve;
   });
 
-  const viewer = Services.viewerForDoc(win.document);
-  viewer.isTrustedViewer().then(isTrusted => {
+  trackImpressionPromise = Services.timerFor(win).timeoutPromise(TIMEOUT_VALUE,
+      promise, 'TrackImpressionPromise timeout').catch(error => {
+        dev().warn('IMPRESSION', error);
+      });
+
+  Services.viewerForDoc(win.document).isTrustedViewer().then(isTrusted => {
     // Currently this feature is launched for trusted viewer, but still
     // experiment guarded for all AMP docs.
     if (!isTrusted && !isExperimentOn(win, 'alp')) {
       resolveImpression();
       return;
     }
 
-    trackImpressionFromClickParam(win, viewer, resolveImpression);
+    const replaceUrlPromise = handleReplaceUrl(win);
+    const clickUrlPromise = handleClickUrl(win);
+
+    Promise.all([replaceUrlPromise, clickUrlPromise]).then(() => {
+      resolveImpression();
+    }, () => {});
   });
 }
 
+/**
+ * Signal that impression tracking is not relevant in this environment.
+ */
+export function doNotTrackImpression() {
+  trackImpressionPromise = Promise.resolve();
+}
+
+/**
+ * Handle the getReplaceUrl and return a promise when url is replaced
+ * Only handles replaceUrl when viewer indicates AMP to do so. Viewer should indicate
+ * by setting the legacy replaceUrl init param and add `replaceUrl` to its capability param.
+ * Future plan is to change the type of legacy init replaceUrl param from url string
+ * to boolean value.
+ * Please NOTE replaceUrl and adLocation will never arrive at same time,
+ * so there is no race condition on the order of handling url replacement.
+ * @param {!Window} win
+ * @return {!Promise}
+ */
+function handleReplaceUrl(win) {
+  const viewer = Services.viewerForDoc(win.document);
+
+  // ReplaceUrl substitution doesn't have to wait until the document is visible
+  if (!viewer.getParam('replaceUrl')) {
+    // The init replaceUrl param serve as a signal on whether replaceUrl is
+    // required for this doc.
+    return Promise.resolve();
+  }
+
+  if (!viewer.hasCapability('replaceUrl')) {
+    // If Viewer is not capability of providing async replaceUrl, use the legacy
+    // init replaceUrl param.
+    viewer.replaceUrl(/** @type {string} */ (viewer.getParam('replaceUrl')));
+    return Promise.resolve();
+  }
+
+  // request async replaceUrl is viewer support getReplaceUrl.
+  return viewer.sendMessageAwaitResponse('getReplaceUrl', undefined)
```

Comment:
why explicit `undefined` as second param? leave empty?

Reply:
The JsonDoc
`{?JsonObject|string|undefined}` doesn't allow me to leave it empty.

Reply:
Document the param with an inline comment? `('getReplaceUrl', /* foo */ undefined)`